libs/remix-ui/workspace/src/lib/actions/workspace.ts

```diff
@@ -358,46 +359,83 @@ export const switchToWorkspace = async (name: string) => {
   }
 }
 
+const loadFile = (name, file, provider, cb?): void => {
+  const fileReader = new FileReader()
+
+  fileReader.onload = async function (event) {
+    if (checkSpecialChars(file.name)) {
+      return dispatch(displayNotification('File Upload Failed', 'Special characters are not allowed', 'Close', null, async () => { }))
+    }
+    try {
+      await provider.set(name, event.target.result)
+    } catch (error) {
+      return dispatch(displayNotification('File Upload Failed', 'Failed to create file ' + name, 'Close', null, async () => { }))
+    }
+
+    const config = plugin.registry.get('config').api
+    const editor = plugin.registry.get('editor').api
+
+    if ((config.get('currentFile') === name) && (editor.currentContent() !== event.target.result)) {
+      editor.setText(name, event.target.result)
+    }
+  }
+  fileReader.readAsText(file)
+  cb && cb(null, true)
+}
+
 export const uploadFile = async (target, targetFolder: string, cb?: (err: Error, result?: string | number | boolean | Record<string, any>) => void) => {
   // TODO The file explorer is merely a view on the current state of
   // the files module. Please ask the user here if they want to overwrite
   // a file and then just use `files.add`. The file explorer will
   // pick that up via the 'fileAdded' event from the files module.
   [...target.files].forEach(async (file) => {
     const workspaceProvider = plugin.fileProviders.workspace
-    const loadFile = (name: string): void => {
-      const fileReader = new FileReader()
-
-      fileReader.onload = async function (event) {
-        if (checkSpecialChars(file.name)) {
-          return dispatch(displayNotification('File Upload Failed', 'Special characters are not allowed', 'Close', null, async () => { }))
-        }
-        try {
-          await workspaceProvider.set(name, event.target.result)
-        } catch (error) {
-          return dispatch(displayNotification('File Upload Failed', 'Failed to create file ' + name, 'Close', null, async () => { }))
-        }
-
-        const config = plugin.registry.get('config').api
-        const editor = plugin.registry.get('editor').api
+    const name = targetFolder === '/' ? file.name : `${targetFolder}/${file.name}`
 
-        if ((config.get('currentFile') === name) && (editor.currentContent() !== event.target.result)) {
-          editor.setText(name, event.target.result)
-        }
+    if (!await workspaceProvider.exists(name)) {
+      loadFile(name, file, workspaceProvider, cb)
+    } else {
+      const modalContent: AppModal = {
+        id: 'overwriteUploadFile',
+        title: 'Confirm overwrite',
+        message: `The file ${name} already exists! Would you like to overwrite it?`,
```

Comment:
- First of all the file name should be in quotes.

- I see a problem with OK/Cancel choices. Why not add "(Number)" and allow to upload. We should offer a solution to the user.

Reply:
First is fine. Second is an existing behaviour. I just refactored this code.

Reply:
first is addressed. Second can come in a different PR

Reply:
created issued for second one: https://github.com/ethereum/remix-project/issues/3456